Sources/CodeGen/CXX/CXXTranspiler.swift

```diff
@@ -61,106 +72,54 @@ public struct CXXTranspiler {
-    module.addTopLevelDecl(
-      CXXFunctionDecl(
-        identifier: identifier,
-        output: output,
-        parameters: cxxParams,
-        body: cxxBody,
-        original: decl))
+    return cxxParams
   }
 
-  /// Translate the function body into a CXX entity.
-  private mutating func emit(funBody body: FunctionDecl.Typed.Body) -> CXXStmt {
+  /// Transpiles a function body to a C++.
+  private func cxx(funBody body: FunctionDecl.Typed.Body) -> CXXScopedBlock {
     switch body {
     case .block(let stmt):
-      return emit(brace: stmt)
+      return cxx(brace: stmt)
 
     case .expr(let expr):
-      let exprBody = CXXReturnStmt(expr: emitR(expr: expr), original: AnyNodeID.TypedNode(expr))
-      return CXXScopedBlock(stmts: [exprBody], original: AnyNodeID.TypedNode(expr))
+      let exprBody = CXXReturnStmt(expr: cxx(expr: expr))
+      return CXXScopedBlock(stmts: [exprBody])
     }
   }
 
-  // MARK: Declarations
-
-  /// Emits the given function declaration into `module`.
-  private mutating func emit(type decl: ProductTypeDecl.Typed, into module: inout CXXModule) {
-    assert(program.isGlobal(decl.id))
-
-    let name = CXXIdentifier(decl.identifier.value)
+  /// Transpiles a product type declaration to a C++ class.
+  private func cxx(type src: ProductTypeDecl.Typed) -> CXXClassDecl {
+    assert(wholeValProgram.isGlobal(src.id))
+    return CXXClassDecl(
+      name: CXXIdentifier(src.identifier.value),
+      members: cxx(classMembers: src.members))
+  }
 
-    // Transpile the class membmers.
+  /// Transpiles product type members to C++.
+  private func cxx<C: Collection>(classMembers src: C) -> [CXXClassDecl.ClassMember]
+  where C.Element == AnyDeclID.TypedNode {
     var cxxMembers: [CXXClassDecl.ClassMember] = []
-    for member in decl.members {
+    for member in src {
       switch member.kind {
       case BindingDecl.self:
-        let bindingDecl = BindingDecl.Typed(member)!
-        // Check if the attribute is static or not.
-        var isStatic = false
-        if bindingDecl.memberModifier != nil {
-          switch bindingDecl.memberModifier!.value {
-          case .static:
-            isStatic = true
-          }
-        }
-        // TODO: visit initializer (bindingDecl.initializer)
-        let cxxInitializer: CXXExpr? = nil
-        // TODO: pattern introducer (let, var, sink, inout)
-        // Visit the name patterns.
-        for (_, name) in bindingDecl.pattern.subpattern.names {
-          let varDecl = name.decl
-          let cxxAttribute = CXXClassAttribute(
-            type: CXXTypeExpr(varDecl.type, ast: program.ast)!,
-            name: CXXIdentifier(varDecl.name),
-            initializer: cxxInitializer,
-            isStatic: isStatic,
-            original: varDecl)
-          cxxMembers.append(.attribute(cxxAttribute))
-        }
+        // One binding can expand into multiple class attributes
+        cxx(productTypeBinding: BindingDecl.Typed(member)!, into: &cxxMembers)
```

Comment:
This use site doesn't make sense, which indicates that `cxx` is the wrong name for this function; it's a noun, which indicates its primary purpose is to return something, but this modifies `cxxMembers`.  `transpile` would be a suitable name here, and the use site would read like a sentence.

Reply:
Alternately, you could solve the naming problem this way:

```suggestion
        cxxMembers += cxx(productTypeBinding: BindingDecl.Typed(member)!)
```

We're nowhere near being at the point of needing to avoid the cost of the intermediate allocations.

Reply:
done